crates/biome_js_analyze/src/semantic_analyzers/nursery/use_for_of.rs

```diff
@@ -0,0 +1,384 @@
+                    return None;
+                }
+            }
+
+            AnyJsExpression::cast(reference.syntax().parent()?)
+        })
+        .collect()
+}
+
+/// Ensure the `for` declaration is correctly initialized
+///
+/// ## Valid example
+/// for (let i = 0; i < array.length; i++)
+///
+/// ## Invalid examples
+/// for (let i = 0, x = 1; i < array.length; i++)
+///
+fn is_initializer_valid(initializer: &AnyJsForInitializer) -> Option<bool> {
+    let initializer_declarations = initializer.as_js_variable_declaration()?.declarators();
+    let initializer = initializer_declarations.first()?.ok()?;
+
+    if initializer_declarations.len() > 1 || !is_zero_initialized(&initializer)? {
+        return None;
+    }
+
+    Some(true)
+}
+
+/// Ensure the `for` test is correctly initialized
+///
+/// ## Valid example
+/// for (let i = 0; i < array.length; i++)
+///
+/// ## Invalid examples
+/// for (let i = 0; NOTI < array.length; i++)
+///
+fn is_test_valid(
+    initializer_binding: &JsIdentifierBinding,
+    test: &AnyJsExpression,
+) -> Option<bool> {
+    let test_binary_expression = test.as_js_binary_expression()?;
+    let left = test_binary_expression.left().ok()?;
+    let identifier_expression = left.as_js_identifier_expression()?;
+
+    if initializer_binding.name_token().ok()?.text_trimmed()
+        != identifier_expression
+            .name()
+            .ok()?
+            .value_token()
+            .ok()?
+            .text_trimmed()
+    {
+        return None;
+    }
+
+    if !is_less_than_length_expression(test_binary_expression)? {
+        return None;
+    }
+
+    Some(true)
+}
+
+/// Ensure the `for` update is correctly initialized
+///
+/// ## Valid example
+/// for (let i = 0; i < array.length; i++)
+///
+/// ## Invalid examples
+/// for (let i = 0; i < array.length; i--)
+///
+/// for (let i = 0; i < array.length; NOTI++)
+///
+fn is_update_valid(
+    initializer_binding: &JsIdentifierBinding,
+    update: AnyJsExpression,
+) -> Option<bool> {
+    let incrementable_like = get_incrementable_like(update)?;
+
+    if initializer_binding.name_token().ok()?.text_trimmed()
+        != incrementable_like.get_name_token()?.text_trimmed()
+    {
+        return None;
+    }
+
+    if !incrementable_like.is_increment_by_one()? {
+        return None;
+    }
+
+    Some(true)
+}
+
+fn reference_being_used_by_array(
+    expression: &AnyJsExpression,
+    array_used_in_for: &AnyJsExpression,
+) -> Option<bool> {
+    match expression.parent::<AnyJsExpression>()? {
+        AnyJsExpression::JsComputedMemberExpression(computed_member) => Some(is_node_equal(
+            computed_member.object().ok()?.syntax(),
+            array_used_in_for.syntax(),
+        )),
+        _ => Some(false),
+    }
+}
+
+fn is_delete(expression: &AnyJsExpression, array_used_in_for: &AnyJsExpression) -> Option<bool> {
+    let parent = expression.parent::<AnyJsExpression>()?;
```

Comment:
We can use `grand_parent`. Also, let's explain why we use `grand_parent`, by providing some documentation to the function.

Reply:
I think AnyJsExpression don't have a `grand_parent` method

Reply:
`.syntax().grand_parent` has it :)

Reply:
Okay